src/websocket_ev.c

```diff
@@ -0,0 +1,331 @@
+      for (i = 0; i < count; i++)
+	{
+	  item = json_object_array_get_idx(needle, i);
+
+	  if (json_object_get_type(item) == json_type_string)
+	    {
+	      event_type = json_object_get_string(item);
+	      DPRINTF(E_DBG, L_WEB, "notify callback event received: %s\n", event_type);
+
+	      if (0 == strcmp(event_type, "update"))
+		{
+		  *requested_events |= LISTENER_UPDATE;
+		}
+	      else if (0 == strcmp(event_type, "database"))
+		{
+		  *requested_events |= LISTENER_DATABASE;
+		}
+	      else if (0 == strcmp(event_type, "pairing"))
+		{
+		  *requested_events |= LISTENER_PAIRING;
+		}
+	      else if (0 == strcmp(event_type, "spotify"))
+		{
+		  *requested_events |= LISTENER_SPOTIFY;
+		}
+	      else if (0 == strcmp(event_type, "lastfm"))
+		{
+		  *requested_events |= LISTENER_LASTFM;
+		}
+	      else if (0 == strcmp(event_type, "outputs"))
+		{
+		  *requested_events |= LISTENER_SPEAKER;
+		}
+	      else if (0 == strcmp(event_type, "player"))
+		{
+		  *requested_events |= LISTENER_PLAYER;
+		}
+	      else if (0 == strcmp(event_type, "options"))
+		{
+		  *requested_events |= LISTENER_OPTIONS;
+		}
+	      else if (0 == strcmp(event_type, "volume"))
+		{
+		  *requested_events |= LISTENER_VOLUME;
+		}
+	      else if (0 == strcmp(event_type, "queue"))
+		{
+		  *requested_events |= LISTENER_QUEUE;
+		}
+	    }
+	}
+    }
+
+  json_tokener_free(tokener);
+  json_object_put(request);
+
+  return 0;
+}
+
+static void
+on_msg_cb(struct evws_connection *evws, int type, const unsigned char *data, size_t len, void *arg)
+{
+  struct client *self = arg;
+  const char *msg = (const char *)data;
+
+  process_notify_request(&self->requested_events, msg, len);
+}
+
+static void
+on_close_cb(struct evws_connection *evws, void *arg)
+{
+  struct client *client = NULL;
+  struct client *prev = NULL;
+
+  for (client = clients; client && client != arg; client = clients->next)
+    {
+      prev = client;
+    }
+
+  if (client)
+    {
+      if (prev)
+	prev->next = client->next;
+      else
+	clients = client->next;
+
+      free(client);
+    }
+}
+
+static void
+gencb_ws(struct evhttp_request *req, void *arg)
+{
+  struct client *client;
+
+  client = calloc(1, sizeof(*client));
+
+  client->evws = evws_new_session(req, on_msg_cb, client, 0);
+  if (!client->evws)
+    {
+      free(client);
+      return;
+    }
+
+  evws_connection_set_closecb(client->evws, on_close_cb, client);
+  client->next = clients;
+  clients = client;
+}
+
+int
+websocketev_init(struct evhttp *evhttp)
+{
+  int websocket_port = cfg_getint(cfg_getsec(cfg, "general"), "websocket_port");
+
+  if (websocket_port > 0)
+    {
+      DPRINTF(E_DBG, L_WEB,
+	      "Libevent websocket disabled, using libwebsockets instead. Set "
+	      "websocket_port to 0 to enable it.\n");
+      return 0;
+    }
+
+  evhttp_set_cb(evhttp, "/ws", gencb_ws, NULL);
+
+  listener_add(listener_cb, LISTENER_UPDATE | LISTENER_DATABASE | LISTENER_PAIRING | LISTENER_SPOTIFY | LISTENER_LASTFM
```

Comment:
Yes, that sounds like a good idea. Together with your suggestion to introduce a map for the event names, this would allow to implement a generic forwarding of events through the ws connection.

But ... I would like to keep it out of this PR as it is not directly related - if that is OK with you.